tsdb/head_wal.go

```diff
@@ -41,7 +42,7 @@ import (
 	"github.com/prometheus/prometheus/tsdb/wal"
 )
 
-func (h *Head) loadWAL(r *wal.Reader, multiRef map[chunks.HeadSeriesRef]chunks.HeadSeriesRef, mmappedChunks map[chunks.HeadSeriesRef][]*mmappedChunk) (err error) {
+func (h *Head) loadWAL(ctx context.Context, r *wal.Reader, multiRef map[chunks.HeadSeriesRef]chunks.HeadSeriesRef, mmappedChunks map[chunks.HeadSeriesRef][]*mmappedChunk) (err error) {
```

Comment:
We need to also close the goroutine that is decoding the WAL and not leak the goroutine. You might want to send an empty record in the decoded channel when context is cancelled so that the consumer of the records below is not stuck waiting for another record and can exit.

Reply:
This should be addressed too, I've also added a test for the whole thing that did initially complained about in-flight goroutines, it no longer does